File: exp/c++/Memory.cxx
#include <unistd.h>
#include <iostream>
#include <queue>

struct Request {
  long num;
  char s[1000000];
};

Request* createRequest(long num) {
  Request* r = new Request();
  r->num = num;
  return r;
}

int main (int argc, char *argv[]) {
  if (argc != 4) {
    fprintf(stdout,"Usage: %s numIterations bufferSize cleanUp\n",argv[0]);
    return 1;
  }

  int numIterations = atoi(argv[1]);
  int bufferSize = atoi(argv[2]);
  int cleanUp = atoi(argv[3]);
  fprintf(stderr, "numIterations=%d, bufferSize=%d, cleanUp=%d\n", numIterations, bufferSize, cleanUp);

  std::queue<Request*> requests;
  for (int i = 0; i < bufferSize; ++i) {
    requests.push(createRequest(i));
  }

  for (int i = 0; i < numIterations; ++i) {
    Request* request = requests.front();
    if (cleanUp) delete request;
    requests.pop();

    requests.push(createRequest(i));
  }

  exit(0);
}

File: exp/java/Memory.java
import java.util.LinkedList;

class Request {
  long num;
  byte[] s;

  public Request(long num) {
    this.num = num;
    this.s = new byte[1000000];
  }
}

public class Memory {
  public static void main(String[] args) {
    int numIterations = Integer.parseInt(args[0]);
    int bufferSize = Integer.parseInt(args[1]);

    LinkedList<Request> requests = new LinkedList<>();
    for (int i = 0; i < bufferSize; ++i) {
      requests.push(new Request(i));
    }

    for (int i = 0; i < numIterations; ++i) {
      Request request = requests.remove();
      requests.push(new Request(i));
    }
  }
}
